File: problem-solving/_lc-0072-medium-edit-distance/1-top-down.go
/*
https://leetcode.com/problems/edit-distance/

Complexity
----------
Time: O(m*n)
Space: O(m*n)

m and n are lengths of 2 strings.
*/

func minDistance(word1 string, word2 string) int {
    cache := make(map[string]int)
    var F func(int, int) int
    F = func(i, j int) int {
        if i == len(word1) {
            return len(word2) - j
        }
        if j == len(word2) {
            return len(word1) - i
        }
        key := fmt.Sprintf("%d-%d", i, j)
        if v, ok := cache[key]; ok {
            return v
        }
        if word1[i] == word2[j] {
            cache[key] = F(i + 1, j + 1)
        } else {
            cache[key] = 1 + min(
                F(i + 1, j),      // Delete
                F(i, j + 1),      // Insert
                F(i + 1, j + 1),  // Replace
            )
        }
        return cache[key]
    }
    return F(0, 0)
}

func min(x, y, z int) int {
    if x < y && x < z {
        return x
    } else if y < x && y < z {
        return y
    }
    return z
}

File: problem-solving/lc-72-medium-edit-distance/1-top-down.py
"""
https://leetcode.com/problems/edit-distance/

Complexity
----------
Time: O(m*n)
Space: O(m*n)

m and n are lengths of 2 strings.
"""

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        cache = {}
        def distance(i, j):
            if i == len(word1):
                return len(word2) - j
            if j == len(word2):
                return len(word1) - i
            if (i, j) in cache:
                return cache[(i, j)]
            if word1[i] == word2[j]:
                cache[(i, j)] = distance(i + 1, j + 1)
            else:
                cache[(i, j)] = 1 + min(
                    distance(i + 1, j),     # Delete
                    distance(i, j + 1),     # Insert
                    distance(i + 1, j + 1)  # Replace
                    )
            return cache[(i, j)]
        return distance(0, 0)
